plugins/inputs/mesos/mesos.go

```diff
@@ -593,11 +672,119 @@ func (m *Mesos) gatherMainMetrics(u *url.URL, role Role, acc telegraf.Accumulato
 		}
 	}
 
+	var includeTaggedFields bool
+	for _, col := range m.MasterCols {
+		if col == "framework_offers" || col == "allocator" {
+			includeTaggedFields = true
+			break
+		}
+	}
+
+	taggedFields := map[string][]TaggedField{}
+	extraTags := map[string]map[string]string{}
+
+	if includeTaggedFields {
+		for metricName, val := range jf.Fields {
+			if !strings.HasPrefix(metricName, "master/frameworks/") && !strings.HasPrefix(metricName, "allocator/") {
+				continue
+			}
+
+			parts := strings.Split(metricName, "/")
+			if (parts[0] == "master" && len(parts) < 5) || (parts[0] == "allocator" && len(parts) <= 5) {
+				continue
+			}
+
+			handleTaggedFields(parts, val, taggedFields, extraTags)
+			delete(jf.Fields, metricName)
+		}
+	}
+
 	acc.AddFields("mesos", jf.Fields, tags)
 
+	for tfh, tfs := range taggedFields {
+		fields := map[string]interface{}{}
+		for _, tf := range tfs {
+			fields[tf.FieldName] = tf.Value
+		}
+		for k, v := range tags {
+			extraTags[tfh][k] = v
+		}
+
+		acc.AddFields("mesos", fields, extraTags[tfh])
+	}
+
 	return nil
 }
 
+func handleTaggedFields(parts []string, val interface{}, taggedFields map[string][]TaggedField, extraTags map[string]map[string]string) {
```

Comment:
I think this function should return `TaggedField` so it won't require `taggedFields` and `extraTags`. I prefer to mutate state in single function and use helper functions to return value then modify state.